graphene_sqlalchemy/registry.py

```diff
@@ -1,22 +1,83 @@
+from collections import OrderedDict
+
+from sqlalchemy.types import Enum as SQLAlchemyEnumType
+
+from graphene import Enum
+
+from .utils import to_enum_value_name, to_type_name
+
+
 class Registry(object):
-    def __init__(self):
+    def __init__(self, check_duplicate_registration=False):
+        self.check_duplicate_registration = check_duplicate_registration
         self._registry = {}
         self._registry_models = {}
         self._registry_composites = {}
+        self._registry_enums = {}
+        self._registry_sort_params = {}
 
     def register(self, cls):
         from .types import SQLAlchemyObjectType
 
-        assert issubclass(cls, SQLAlchemyObjectType), (
-            "Only classes of type SQLAlchemyObjectType can be registered, "
-            'received "{}"'
-        ).format(cls.__name__)
-        assert cls._meta.registry == self, "Registry for a Model have to match."
-        # assert self.get_type_for_model(cls._meta.model) in [None, cls], (
-        #     'SQLAlchemy model "{}" already associated with '
-        #     'another type "{}".'
-        # ).format(cls._meta.model, self._registry[cls._meta.model])
-        self._registry[cls._meta.model] = cls
+        if not issubclass(cls, SQLAlchemyObjectType):
+            raise TypeError(
+                "Only classes of type SQLAlchemyObjectType can be registered, "
+                'received "{}"'.format(cls.__name__)
+            )
+        if cls._meta.registry != self:
+            raise TypeError("Registry for a Model have to match.")
+
+        registered_cls = (
+            self._registry.get(cls._meta.model)
+            if self.check_duplicate_registration
+            else None
+        )
+        if registered_cls:
+            if cls != registered_cls:
+                raise TypeError(
+                    "Different object types registered for the same model {}:"
+                    " tried to register {}, but {} existed already.".format(
+                        cls._meta.model, cls, registered_cls
+                    )
+                )
+        else:
+            self._registry[cls._meta.model] = cls
+
+    def register_enum(self, name, members):
```

Comment:
Because not all enums are derived from SQLAlchemyEnumTypes. Some are created for sorting purposes, they don't exist as SQLAlchemyEnumTypes. All of the enums should use the same registry because Enum names in the schema must be unique -  you want to get an error when a sort enum accidentally has the same name as an existing enum for a column, and you want to have a way for requesting registered enums by name.

Reply:
Right. I will think about the whole mechanism again and try to create a better-thought-out solution. But maybe I'll do this as a new PR.